// styled-minimal/lib/Alert.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Alert = exports.StyledAlert = void 0;
var React = require("react");
var styled_components_1 = require("styled-components");
var helpers_1 = require("./utils/helpers");
var system_1 = require("./utils/system");
var MutationWrapper_1 = require("./MutationWrapper");
exports.StyledAlert = styled_components_1.default('div')(function (props) {
    var _a = helpers_1.getTheme(props, 'alert'), borderRadius = _a.borderRadius, maxWidth = _a.maxWidth, padding = _a.padding;
    return styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      ", ";\n      ", ";\n      border-radius: ", ";\n      font-size: inherit;\n      line-height: ", ";\n      max-width: ", ";\n      padding: ", ";\n      width: 100%;\n\n      a {\n        ", ";\n      }\n    "], ["\n      ", ";\n      ", ";\n      border-radius: ", ";\n      font-size: inherit;\n      line-height: ", ";\n      max-width: ", ";\n      padding: ", ";\n      width: 100%;\n\n      a {\n        ", ";\n      }\n    "])), system_1.baseStyles, system_1.variantStyles, helpers_1.px(borderRadius), system_1.lineHeightStyles, helpers_1.px(maxWidth), padding, system_1.colorStyles);
}, system_1.mutationStyles(), system_1.sx, system_1.system);
exports.Alert = React.forwardRef(function (props, ref) { return (React.createElement(MutationWrapper_1.default, __assign({}, props),
    React.createElement(exports.StyledAlert, __assign({ ref: ref }, props)))); });
exports.Alert.defaultProps = {
    dark: false,
    invert: false,
    role: 'alert',
    variant: 'primary',
};
var templateObject_1;
