// styled-minimal/lib/Badge.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Badge = exports.StyledBadge = void 0;
var React = require("react");
var styled_components_1 = require("styled-components");
var helpers_1 = require("./utils/helpers");
var system_1 = require("./utils/system");
exports.StyledBadge = styled_components_1.default('sup')(function (props) {
    var dot = props.dot;
    var _a = helpers_1.getTheme(props, 'badge'), borderRadius = _a.borderRadius, fontSize = _a.fontSize, fontWeight = _a.fontWeight, lineHeight = _a.lineHeight, padding = _a.padding, size = _a.size, sizeDot = _a.sizeDot;
    return styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      ", ";\n      ", ";\n      border-radius: ", ";\n      display: inline-flex;\n      font-size: ", ";\n      font-weight: ", ";\n      justify-content: center;\n      line-height: ", ";\n      height: ", ";\n      padding: ", ";\n      min-width: ", ";\n    "], ["\n      ", ";\n      ", ";\n      border-radius: ", ";\n      display: inline-flex;\n      font-size: ", ";\n      font-weight: ", ";\n      justify-content: center;\n      line-height: ", ";\n      height: ", ";\n      padding: ", ";\n      min-width: ", ";\n    "])), system_1.baseStyles, system_1.variantStyles, dot ? '100%' : helpers_1.px(borderRadius), dot ? 0 : helpers_1.px(fontSize), fontWeight, dot ? 0 : helpers_1.px(lineHeight), helpers_1.px(dot ? sizeDot : size), dot ? 0 : "0 " + helpers_1.px(padding), helpers_1.px(dot ? sizeDot : size));
}, system_1.sx, system_1.system);
exports.Badge = React.forwardRef(function (props, ref) { return (React.createElement(exports.StyledBadge, __assign({ ref: ref }, props))); });
exports.Badge.defaultProps = {
    dark: false,
    dot: false,
    invert: false,
    variant: 'primary',
};
var templateObject_1;
