// styled-minimal/lib/Button.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Button = exports.StyledButton = void 0;
var React = require("react");
var styled_components_1 = require("styled-components");
var helpers_1 = require("./utils/helpers");
var system_1 = require("./utils/system");
var Badge_1 = require("./Badge");
var MutationWrapper_1 = require("./MutationWrapper");
exports.StyledButton = styled_components_1.default('button')(function (props) {
    var block = props.block, busy = props.busy, invert = props.invert, _a = props.size, size = _a === void 0 ? 'md' : _a;
    var _b = helpers_1.getTheme(props, 'button'), borderRadius = _b.borderRadius, disabledOpacity = _b.disabledOpacity, fontSize = _b.fontSize, lineHeight = _b.lineHeight, loader = _b.loader, padding = _b.padding;
    return styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      ", ";\n      ", ";\n      ", ";\n      ", ";\n      align-items: center;\n      border-radius: ", ";\n      box-shadow: none;\n      cursor: pointer;\n      display: inline-flex;\n      font-family: inherit;\n      font-size: ", ";\n      line-height: ", ";\n      padding: ", " ", ";\n      text-decoration: none;\n      width: ", ";\n\n      &:disabled {\n        opacity: ", ";\n        pointer-events: none;\n      }\n\n      &:focus {\n        outline-color: ", ";\n      }\n\n      ", " {\n        margin-left: 5px;\n      }\n    "], ["\n      ", ";\n      ", ";\n      ", ";\n      ", ";\n      align-items: center;\n      border-radius: ", ";\n      box-shadow: none;\n      cursor: pointer;\n      display: inline-flex;\n      font-family: inherit;\n      font-size: ", ";\n      line-height: ", ";\n      padding: ", " ", ";\n      text-decoration: none;\n      width: ", ";\n\n      &:disabled {\n        opacity: ", ";\n        pointer-events: none;\n      }\n\n      &:focus {\n        outline-color: ", ";\n      }\n\n      ", " {\n        margin-left: 5px;\n      }\n    "])), system_1.baseStyles, system_1.appearanceStyles, system_1.variantStyles, busy ? loader(invert ? '#ccc' : '#fff') : '', helpers_1.px(borderRadius[size]), helpers_1.px(fontSize[size]), lineHeight, helpers_1.px(padding[size][0]), helpers_1.px(padding[size][1]), block ? '100%' : 'auto', disabledOpacity, helpers_1.getColor, Badge_1.StyledBadge);
}, system_1.mutationStyles(), system_1.sx, system_1.system);
exports.Button = React.forwardRef(function (props, ref) {
    var block = props.block, invert = props.invert;
    var button = React.createElement(exports.StyledButton, __assign({ ref: ref }, props));
    if (invert) {
        return (React.createElement(MutationWrapper_1.default, __assign({ display: block ? 'block' : 'inline-flex' }, props), button));
    }
    return button;
});
exports.Button.defaultProps = {
    block: false,
    busy: false,
    dark: false,
    disabled: false,
    invert: false,
    size: 'md',
    type: 'button',
    variant: 'primary',
};
var templateObject_1;
