// styled-minimal/lib/Container.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Container = exports.StyledContainer = void 0;
var React = require("react");
var styled_components_1 = require("styled-components");
var helpers_1 = require("./utils/helpers");
var system_1 = require("./utils/system");
exports.StyledContainer = styled_components_1.default('div')(function (props) {
    var ySpacing = props.ySpacing, fullScreen = props.fullScreen;
    var ySpacingStyles;
    var fullScreenStyles;
    if (ySpacing) {
        ySpacingStyles = styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n        padding-bottom: ", ";\n        padding-top: ", ";\n\n        ", ";\n      "], ["\n        padding-bottom: ", ";\n        padding-top: ", ";\n\n        ",
            ";\n      "])), helpers_1.spacer(3), helpers_1.spacer(3), helpers_1.responsive({
            lg: {
                paddingBottom: helpers_1.spacer(4)(props),
                paddingTop: helpers_1.spacer(4)(props),
            },
        }));
    }
    if (fullScreen) {
        fullScreenStyles = styled_components_1.css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n        align-items: center;\n        display: flex;\n        flex-direction: column;\n        justify-content: center;\n        min-height: 100vh;\n      "], ["\n        align-items: center;\n        display: flex;\n        flex-direction: column;\n        justify-content: center;\n        min-height: 100vh;\n      "])));
    }
    return styled_components_1.css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n      ", ";\n      margin-left: auto;\n      margin-right: auto;\n      padding-left: ", ";\n      padding-right: ", ";\n      position: relative;\n      width: 100%;\n      ", ";\n      ", ";\n      ", ";\n    "], ["\n      ", ";\n      margin-left: auto;\n      margin-right: auto;\n      padding-left: ", ";\n      padding-right: ", ";\n      position: relative;\n      width: 100%;\n      ", ";\n      ", ";\n      ",
        ";\n    "])), system_1.baseStyles, helpers_1.spacer(3), helpers_1.spacer(3), fullScreenStyles, ySpacingStyles, helpers_1.responsive({
        lg: {
            paddingLeft: helpers_1.spacer(4)(props),
            paddingRight: helpers_1.spacer(4)(props),
        },
    }));
}, system_1.sx, system_1.system);
exports.Container = React.forwardRef(function (props, ref) { return (React.createElement(exports.StyledContainer, __assign({ ref: ref }, props))); });
exports.Container.defaultProps = {
    fullScreen: false,
    ySpacing: false,
};
var templateObject_1, templateObject_2, templateObject_3;
