// styled-minimal/lib/RadioGroup.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RadioGroup = void 0;
var React = require("react");
var FormGroup_1 = require("./FormGroup");
var Input_1 = require("./Input");
var RadioGroup = /** @class */ (function (_super) {
    __extends(RadioGroup, _super);
    function RadioGroup(props) {
        var _this = _super.call(this, props) || this;
        _this.update = function (selectedValue) {
            _this.setState({ selectedValue: selectedValue });
        };
        _this.handleChange = function (_a) {
            var target = _a.target;
            var _b = _this.props, value = _b.value, disabled = _b.disabled, onChange = _b.onChange;
            var numericValue = Number(target.value);
            var currentValue = !Number.isNaN(numericValue) ? numericValue : target.value;
            if (onChange) {
                onChange(currentValue);
            }
            if (!value && !disabled) {
                _this.update(currentValue);
            }
        };
        _this.state = {
            selectedValue: props.value || props.defaultValue,
        };
        return _this;
    }
    RadioGroup.prototype.componentDidUpdate = function (prevProps) {
        var value = this.props.value;
        if (value && prevProps.value !== value) {
            this.update(value);
        }
    };
    RadioGroup.prototype.render = function () {
        var _this = this;
        var selectedValue = this.state.selectedValue;
        var _a = this.props, dark = _a.dark, defaultValue = _a.defaultValue, disabled = _a.disabled, inline = _a.inline, invert = _a.invert, name = _a.name, onChange = _a.onChange, options = _a.options, size = _a.size, value = _a.value, variant = _a.variant, rest = __rest(_a, ["dark", "defaultValue", "disabled", "inline", "invert", "name", "onChange", "options", "size", "value", "variant"]);
        if (!options || !options.length) {
            return null;
        }
        return (React.createElement(FormGroup_1.FormGroup, __assign({}, rest, { role: "radiogroup" }), options.map(function (d) { return (React.createElement(Input_1.Radio, { key: d.value, checked: d.value === selectedValue, dark: dark, disabled: disabled || d.disabled, inline: inline, invert: invert, label: d.label || d.value, name: name, onChange: _this.handleChange, size: size, value: d.value, variant: d.variant || variant })); })));
    };
    RadioGroup.defaultProps = {
        disabled: false,
        size: 'md',
        variant: 'primary',
    };
    return RadioGroup;
}(React.PureComponent));
exports.RadioGroup = RadioGroup;
