// styled-minimal/lib/Switch.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Switch = exports.StyledSwitch = void 0;
var React = require("react");
var styled_components_1 = require("styled-components");
var helpers_1 = require("./utils/helpers");
var system_1 = require("./utils/system");
var Label_1 = require("./Label");
var StyledInput = styled_components_1.default('input')(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  bottom: 0;\n  left: 0;\n  opacity: 0.0001;\n  position: absolute;\n  right: 0;\n  top: 0;\n"], ["\n  bottom: 0;\n  left: 0;\n  opacity: 0.0001;\n  position: absolute;\n  right: 0;\n  top: 0;\n"])));
var StyledTrack = styled_components_1.default('span')(function (props) {
    var size = props.size, status = props.status;
    var sizes = helpers_1.getTheme(props, 'switchOptions').sizes;
    var color = status ? helpers_1.getColor(props) : '#ccc';
    return styled_components_1.css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    background-color: ", ";\n    border-radius: ", ";\n    bottom: 0;\n    left: 0;\n    position: absolute;\n    right: 0;\n    top: 0;\n  "], ["\n    background-color: ", ";\n    border-radius: ", ";\n    bottom: 0;\n    left: 0;\n    position: absolute;\n    right: 0;\n    top: 0;\n  "])), color, helpers_1.px(sizes[size].borderRadius));
});
var StyledButton = styled_components_1.default('span')(function (props) {
    var disabled = props.disabled, size = props.size, status = props.status;
    var _a = helpers_1.getTheme(props, 'switchOptions'), innerBorderRadius = _a.innerBorderRadius, sizes = _a.sizes;
    var currentSize = sizes[size];
    return styled_components_1.css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n    background-color: #fff;\n    border-radius: ", ";\n    bottom: ", ";\n    left: ", ";\n    opacity: ", ";\n    position: absolute;\n    top: ", ";\n    transition: left 0.1s ease;\n    width: ", ";\n  "], ["\n    background-color: #fff;\n    border-radius: ", ";\n    bottom: ", ";\n    left: ", ";\n    opacity: ", ";\n    position: absolute;\n    top: ", ";\n    transition: left 0.1s ease;\n    width: ", ";\n  "])), helpers_1.px(innerBorderRadius), helpers_1.px(currentSize.space), status ? '50%' : helpers_1.px(currentSize.space), disabled ? 0.7 : 1, helpers_1.px(currentSize.space), helpers_1.px(currentSize.height - currentSize.space * 2));
});
exports.StyledSwitch = styled_components_1.default('div')(function (props) {
    var disabled = props.disabled, label = props.label, size = props.size;
    var _a = helpers_1.getTheme(props, 'switchOptions'), marginLabel = _a.marginLabel, sizes = _a.sizes;
    var currentSize = sizes[size];
    return styled_components_1.css(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n      ", ";\n      cursor: ", ";\n      height: ", ";\n      margin-right: ", ";\n      opacity: ", ";\n      position: relative;\n      user-select: none;\n      vertical-align: middle;\n      width: ", ";\n    "], ["\n      ", ";\n      cursor: ", ";\n      height: ", ";\n      margin-right: ", ";\n      opacity: ", ";\n      position: relative;\n      user-select: none;\n      vertical-align: middle;\n      width: ", ";\n    "])), system_1.baseStyles, disabled ? 'default' : 'pointer', helpers_1.px(currentSize.height), label ? helpers_1.px(marginLabel) : 0, disabled ? 0.8 : 1, helpers_1.px(currentSize.width));
}, system_1.sx, system_1.system);
var SwitchClass = /** @class */ (function (_super) {
    __extends(SwitchClass, _super);
    function SwitchClass(props) {
        var _this = _super.call(this, props) || this;
        _this.update = function (status) {
            _this.setState({ status: status });
        };
        _this.handleChange = function (e) {
            var checked = e.target.checked;
            var onChange = _this.props.onChange;
            _this.update(checked);
            /* istanbul ignore else */
            if (onChange) {
                onChange(checked);
            }
        };
        _this.state = {
            status: helpers_1.isBoolean(props.checked) ? props.checked : props.defaultChecked,
        };
        return _this;
    }
    SwitchClass.prototype.componentDidUpdate = function (prevProps) {
        var checked = this.props.checked;
        if (helpers_1.isBoolean(checked) && prevProps.checked !== checked) {
            this.update(checked);
        }
    };
    SwitchClass.prototype.render = function () {
        var status = this.state.status;
        var _a = this.props, checked = _a.checked, defaultChecked = _a.defaultChecked, disabled = _a.disabled, innerRef = _a.innerRef, label = _a.label, name = _a.name, size = _a.size, variant = _a.variant, rest = __rest(_a, ["checked", "defaultChecked", "disabled", "innerRef", "label", "name", "size", "variant"]);
        var value = status ? 'on' : 'off';
        var checkStatus = {};
        var handleChange;
        if (helpers_1.isBoolean(checked)) {
            checkStatus.checked = checked;
        }
        else {
            handleChange = this.handleChange;
            checkStatus.defaultChecked = defaultChecked;
        }
        return (React.createElement(Label_1.Label, { cursor: disabled ? 'default' : 'pointer' },
            React.createElement(StyledInput, __assign({ "aria-checked": status, "aria-label": !label ? name : undefined, disabled: disabled || helpers_1.isBoolean(checked), name: name, onChange: handleChange, ref: innerRef, role: "switch", type: "checkbox", value: value }, checkStatus)),
            React.createElement(exports.StyledSwitch, __assign({ disabled: disabled, label: label, size: size, status: status }, rest),
                React.createElement(StyledTrack, { size: size, status: status, variant: variant }),
                React.createElement(StyledButton, { disabled: disabled, size: size, status: status, variant: variant })),
            label));
    };
    SwitchClass.defaultProps = {
        defaultChecked: false,
        disabled: false,
        size: 'md',
        variant: 'primary',
    };
    return SwitchClass;
}(React.PureComponent));
exports.Switch = React.forwardRef(function (props, ref) { return (React.createElement(SwitchClass, __assign({}, props, { innerRef: ref }))); });
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
