// styled-minimal/lib/Table.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Table = exports.StyledTable = void 0;
var React = require("react");
var styled_components_1 = require("styled-components");
var helpers_1 = require("./utils/helpers");
var system_1 = require("./utils/system");
var stylesCaption = function (props) {
    var _a = helpers_1.getTheme(props, 'table'), captionColor = _a.captionColor, captionPadding = _a.captionPadding;
    return styled_components_1.css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    caption-side: bottom;\n    color: ", ";\n    font-size: 90%;\n    padding-bottom: ", ";\n    padding-top: ", ";\n    text-align: left;\n  "], ["\n    caption-side: bottom;\n    color: ", ";\n    font-size: 90%;\n    padding-bottom: ", ";\n    padding-top: ", ";\n    text-align: left;\n  "])), captionColor, helpers_1.px(captionPadding), helpers_1.px(captionPadding));
};
var stylesCellBorder = function (props) {
    var frame = props.frame, clean = props.clean, dark = props.dark;
    var borderColors = helpers_1.getTheme(props, 'table').borderColors;
    if (clean) {
        return '';
    }
    return styled_components_1.css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    ", ": 1px solid ", ";\n  "], ["\n    ", ": 1px solid ", ";\n  "])), frame ? 'border' : 'border-top', borderColors[dark ? 'secondary' : 'primary']);
};
var stylesHeadBackgroundColor = function (props) {
    var head = props.head;
    var headColors = helpers_1.getTheme(props, 'table').headColors;
    return styled_components_1.css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n    background-color: ", ";\n  "], ["\n    background-color: ", ";\n  "])), headColors[head] || 'transparent');
};
var stylesHeadCellBorder = function (props) {
    var frame = props.frame, clean = props.clean, dark = props.dark;
    var borderColors = helpers_1.getTheme(props, 'table').borderColors;
    var colorProp = borderColors[dark ? 'secondary' : 'primary'];
    if (clean) {
        return '';
    }
    return styled_components_1.css(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n    ", ";\n    ", ";\n  "], ["\n    ", ";\n    ", ";\n  "])), frame ? "border: 1px solid " + colorProp : "border-bottom: 2px solid " + colorProp, frame && 'border-bottom-width: 2px');
};
var stylesHeadColor = function (props) {
    var head = props.head;
    var colors = helpers_1.getTheme(props, 'table').colors;
    if (head !== 'normal') {
        return styled_components_1.css(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n      color: ", ";\n    "], ["\n      color: ", ";\n    "])), colors[head === 'dark' ? 'primary' : 'secondary']);
    }
    return '';
};
var stylesPadding = function (props) {
    var _a = props.size, size = _a === void 0 ? 'md' : _a;
    var padding = helpers_1.getTheme(props, 'table').padding;
    return styled_components_1.css(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n    padding: ", ";\n  "], ["\n    padding: ", ";\n  "])), helpers_1.px(padding[size]));
};
var stylesStriped = function (props) {
    var dark = props.dark, striped = props.striped;
    var stripedColors = helpers_1.getTheme(props, 'table').stripedColors;
    if (striped) {
        return styled_components_1.css(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n      background-color: ", ";\n    "], ["\n      background-color: ", ";\n    "])), stripedColors[dark ? 'secondary' : 'primary']);
    }
    return '';
};
exports.StyledTable = styled_components_1.default('table')(function (props) {
    var frame = props.frame, clean = props.clean, dark = props.dark;
    var _a = helpers_1.getTheme(props, 'table'), borderColors = _a.borderColors, colors = _a.colors;
    return styled_components_1.css(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n      ", ";\n      background-color: ", ";\n      ", ";\n      border-collapse: collapse;\n      color: ", ";\n      width: 100%;\n\n      th {\n        text-align: inherit;\n      }\n\n      th,\n      td {\n        ", ";\n        ", ";\n        vertical-align: top;\n      }\n\n      thead {\n        ", ";\n        ", ";\n      }\n\n      thead th {\n        ", ";\n        vertical-align: bottom;\n      }\n\n      tbody tr:nth-of-type(odd) {\n        ", ";\n      }\n\n      caption {\n        ", ";\n      }\n    "], ["\n      ", ";\n      background-color: ", ";\n      ",
        ";\n      border-collapse: collapse;\n      color: ", ";\n      width: 100%;\n\n      th {\n        text-align: inherit;\n      }\n\n      th,\n      td {\n        ", ";\n        ", ";\n        vertical-align: top;\n      }\n\n      thead {\n        ", ";\n        ", ";\n      }\n\n      thead th {\n        ", ";\n        vertical-align: bottom;\n      }\n\n      tbody tr:nth-of-type(odd) {\n        ", ";\n      }\n\n      caption {\n        ", ";\n      }\n    "])), system_1.baseStyles, colors[dark ? 'secondary' : 'primary'], frame && !clean
        ? "border: 1px solid " + borderColors[dark ? 'secondary' : 'primary'] + ";"
        : '', colors[dark ? 'primary' : 'secondary'], stylesCellBorder, stylesPadding, stylesHeadBackgroundColor, stylesHeadColor, stylesHeadCellBorder, stylesStriped, stylesCaption);
}, system_1.sx, system_1.system);
exports.Table = React.forwardRef(function (props, ref) { return (React.createElement(exports.StyledTable, __assign({ ref: ref }, props))); });
exports.Table.defaultProps = {
    frame: false,
    clean: false,
    dark: false,
    head: 'normal',
    size: 'md',
    striped: false,
};
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8;
