// styled-minimal/esm/Alert.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import * as React from 'react';
import styled, { css } from 'styled-components';
import { getTheme, px } from './utils/helpers';
import { baseStyles, colorStyles, lineHeightStyles, mutationStyles, sx, system, variantStyles, } from './utils/system';
import MutationWrapper from './MutationWrapper';
export var StyledAlert = styled('div')(function (props) {
    var _a = getTheme(props, 'alert'), borderRadius = _a.borderRadius, maxWidth = _a.maxWidth, padding = _a.padding;
    return css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n      ", ";\n      ", ";\n      border-radius: ", ";\n      font-size: inherit;\n      line-height: ", ";\n      max-width: ", ";\n      padding: ", ";\n      width: 100%;\n\n      a {\n        ", ";\n      }\n    "], ["\n      ", ";\n      ", ";\n      border-radius: ", ";\n      font-size: inherit;\n      line-height: ", ";\n      max-width: ", ";\n      padding: ", ";\n      width: 100%;\n\n      a {\n        ", ";\n      }\n    "])), baseStyles, variantStyles, px(borderRadius), lineHeightStyles, px(maxWidth), padding, colorStyles);
}, mutationStyles(), sx, system);
export var Alert = React.forwardRef(function (props, ref) { return (React.createElement(MutationWrapper, __assign({}, props),
    React.createElement(StyledAlert, __assign({ ref: ref }, props)))); });
Alert.defaultProps = {
    dark: false,
    invert: false,
    role: 'alert',
    variant: 'primary',
};
var templateObject_1;
