// styled-minimal/esm/List.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import * as React from 'react';
import styled, { css } from 'styled-components';
import { getTheme, px } from './utils/helpers';
import { baseStyles, sx, system } from './utils/system';
var stylesItem = function (props) {
    var _a = props.size, size = _a === void 0 ? 'md' : _a;
    var padding = getTheme(props, 'list').padding;
    return css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    padding: ", ";\n  "], ["\n    padding: ", ";\n  "])), px(padding[size]));
};
var stylesDD = function (props) {
    var _a = props.size, size = _a === void 0 ? 'md' : _a;
    var padding = getTheme(props, 'list').padding;
    return css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    margin-left: ", ";\n    padding: 0 ", " ", " 0;\n  "], ["\n    margin-left: ", ";\n    padding: 0 ", " ", " 0;\n  "])), px(padding[size]), px(padding[size]), px(padding[size]));
};
var stylesSibling = function (props) {
    var frame = props.frame;
    var borderColor = getTheme(props, 'list').borderColor;
    if (!frame) {
        return undefined;
    }
    return css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n    border-top: 1px solid ", ";\n  "], ["\n    border-top: 1px solid ", ";\n  "])), borderColor);
};
var stylesChildren = function (props) {
    var as = props.as, frame = props.frame, inline = props.inline, styleType = props.styleType;
    var borderColor = getTheme(props, 'list').borderColor;
    var border = "1px solid " + borderColor;
    if (as === 'dl') {
        return css(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n      dt {\n        font-weight: 700;\n        ", "\n        ", ";\n        ", ";\n        ", ";\n      }\n\n      dd {\n        ", ";\n        ", ";\n        ", ";\n        ", ";\n        ", ";\n\n        + dt {\n          ", ";\n        }\n\n        &:last-of-type {\n          ", ";\n        }\n      }\n\n      dt:last-of-type {\n        ", ";\n      }\n    "], ["\n      dt {\n        font-weight: 700;\n        ", "\n        ", ";\n        ", ";\n        ", ";\n      }\n\n      dd {\n        ", ";\n        ", ";\n        ", ";\n        ", ";\n        ", ";\n\n        + dt {\n          ", ";\n        }\n\n        &:last-of-type {\n          ", ";\n        }\n      }\n\n      dt:last-of-type {\n        ", ";\n      }\n    "])), stylesItem, frame && inline && "border-bottom: " + border + ";", frame && inline && "border-right: " + border + ";", inline && "flex-basis: 30%;", stylesDD, frame && inline && "border-bottom: " + border + ";", inline && stylesItem, inline && "flex-basis: 70%;", inline && "margin-left: 0;", !inline && stylesSibling, inline && "border-bottom: 0;", inline && "border-bottom: 0;");
    }
    return css(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n    > li {\n      ", ";\n      ", ";\n      line-height: 1.2;\n      ", "\n\n      + li {\n        ", "\n      }\n    }\n  "], ["\n    > li {\n      ", ";\n      ", ";\n      line-height: 1.2;\n      ", "\n\n      + li {\n        ", "\n      }\n    }\n  "])), frame && (as === 'ol' || styleType !== 'none' || inline) && "border-left: " + border + ";", inline && 'flex: 1', stylesItem, stylesSibling);
};
export var StyledList = styled('ul')(function (props) {
    var as = props.as, frame = props.frame, inline = props.inline, styleType = props.styleType;
    var _a = getTheme(props, 'list'), borderColor = _a.borderColor, borderRadius = _a.borderRadius;
    var border = "1px solid " + borderColor;
    return css(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n      ", ";\n      ", ";\n      ", ";\n      display: flex;\n      flex-direction: ", ";\n      flex-wrap: ", ";\n      margin: 0;\n      ", ";\n      ", ";\n\n      ", ";\n    "], ["\n      ", ";\n      ", ";\n      ", ";\n      display: flex;\n      flex-direction: ", ";\n      flex-wrap: ", ";\n      margin: 0;\n      ", ";\n      ", ";\n\n      ", ";\n    "])), baseStyles, frame && "border: " + border, frame && "border-radius: " + px(borderRadius), inline ? 'row' : 'column', as === 'dl' && inline ? 'wrap' : 'nowrap', as === 'ul' && styleType === 'none' ? 'padding: 0;' : '', as === 'ul' ? "list-style-type: " + styleType + ";" : '', stylesChildren);
}, sx, system);
export var DL = React.forwardRef(function (props, ref) { return (React.createElement(StyledList, __assign({ as: "dl", ref: ref }, props))); });
DL.defaultProps = {
    frame: true,
    inline: false,
    size: 'md',
};
export var OL = React.forwardRef(function (props, ref) { return (React.createElement(StyledList, __assign({ as: "ol", ref: ref }, props))); });
OL.defaultProps = {
    frame: false,
    inline: false,
    size: 'md',
};
export var UL = React.forwardRef(function (props, ref) { return (React.createElement(StyledList, __assign({ as: "ul", ref: ref }, props))); });
UL.defaultProps = {
    frame: true,
    inline: false,
    size: 'md',
    styleType: 'none',
};
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
