// styled-minimal/src/Textarea.tsx
import * as React from 'react';
import styled, { css } from 'styled-components';

import { getTheme, px } from './utils/helpers';
import {
  baseStyles,
  appearanceStyles,
  inputStyles,
  mutationStyles,
  sx,
  system,
} from './utils/system';

import MutationWrapper from './MutationWrapper';

import { TextareaProps } from './types';

export const StyledTextarea = styled('textarea').attrs(({ id }) => ({
  id,
}))<TextareaProps>(
  (props) => {
    const {
      backgroundColor,
      borderColor,
      borderRadius,
      borderWidth,
      color,
      fontSize,
      lineHeight,
      padding,
      validation,
    } = getTheme(props, 'textarea');
    const { size = 'md', valid } = props;

    let currentBorderColor = borderColor;
    if (valid) {
      currentBorderColor = validation.valid;
    } else if (valid === false) {
      currentBorderColor = validation.invalid;
    }

    return css`
      ${baseStyles};
      ${appearanceStyles};
      background-color: ${backgroundColor};
      border: ${`${px(borderWidth)} solid ${currentBorderColor}`};
      border-radius: ${px(borderRadius)};
      color: ${color};
      display: block;
      font-family: inherit;
      font-size: ${px(fontSize[size])};
      line-height: ${lineHeight};
      margin: 0;
      padding: ${px(padding[size])};
      width: 100%;
    `;
  },
  mutationStyles(),
  inputStyles,
  sx,
  system,
);

export const Textarea = React.forwardRef<HTMLTextAreaElement, TextareaProps>((props, ref) => (
  <MutationWrapper {...props}>
    <StyledTextarea ref={ref} {...props} />
  </MutationWrapper>
));

Textarea.defaultProps = {
  rows: 3,
  size: 'md',
};
