// styled-minimal/esm/utils/system.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { css, } from 'styled-components';
import systemCSS from '@styled-system/css';
import { getColor, getDarkColor, getAccentColor, getMutation, getProps, getTextColor, getTheme, isDefined, px, responsive, } from './helpers';
import { Imperfection } from './imperfection';
import { placeholder } from './mixins';
import { darken, desaturate, lighten } from './polished';
import { styleSystem } from './styles';
export var inputTextTypes = [
    'date',
    'email',
    'file',
    'number',
    'password',
    'search',
    'tel',
    'text',
];
export var appearanceStyles = {
    appearance: 'none',
};
export var baseStyles = {
    boxSizing: 'border-box',
};
export function colorStyles(props) {
    var dark = props.dark, invert = props.invert;
    return css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    color: ", ";\n  "], ["\n    color: ", ";\n  "])), invert || dark ? getColor(props) : getTextColor(props));
}
export function lineHeightStyles(props) {
    return getTheme(props, 'lineHeight');
}
export function mutationStyles(options) {
    var _a = (options || {}).skipTheme, skipTheme = _a === void 0 ? false : _a;
    return function (props) {
        var _a = getMutation(props, skipTheme), complexity = _a.complexity, height = _a.height, mutation = _a.mutation, seed = _a.seed, width = _a.width;
        if (!mutation || props.type === 'color') {
            return css(templateObject_2 || (templateObject_2 = __makeTemplateObject([""], [""])));
        }
        var shape = new Imperfection(seed || '', width, height);
        var path = shape.getPath(mutation, complexity !== null && complexity !== void 0 ? complexity : 0, true);
        return css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n      -webkit-appearance: none;\n      -webkit-clip-path: polygon(", ");\n      border: 0;\n      clip-path: polygon(", ");\n      transition: clip-path 0.5s;\n    "], ["\n      -webkit-appearance: none;\n      -webkit-clip-path: polygon(", ");\n      border: 0;\n      clip-path: polygon(", ");\n      transition: clip-path 0.5s;\n    "])), path, path);
    };
}
export function inputStyles(props) {
    var invert = props.invert, multiple = props.multiple, outline = props.outline, outlineColor = props.outlineColor, _a = props.outlineOffset, outlineOffset = _a === void 0 ? 1 : _a, outlineStyle = props.outlineStyle, outlineWidth = props.outlineWidth, type = props.type, valid = props.valid;
    var elementType = 'input';
    if (isDefined(multiple)) {
        elementType = 'select';
    }
    else if (!isDefined(type)) {
        elementType = 'textarea';
    }
    var _b = getTheme(props, elementType), backgroundColor = _b.backgroundColor, borderColor = _b.borderColor, checkedBgColor = _b.checkedBgColor, color = _b.color, disabledColor = _b.disabledColor, requiredColor = _b.requiredColor, validation = _b.validation;
    var primaryColor = getColor(props);
    var currentBgColor = backgroundColor;
    var currentColor = color;
    var currentBorderColor = borderColor;
    var multipleCheckedBgColor = checkedBgColor || getAccentColor(primaryColor, true);
    if (valid) {
        currentBorderColor = validation.valid;
    }
    else if (valid === false) {
        currentBorderColor = validation.invalid;
    }
    return css(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n    ", ";\n\n    &:focus {\n      ", ";\n      outline-color: ", ";\n      ", ";\n      ", ";\n      ", ";\n    }\n\n    &[multiple] {\n      > option {\n        &:not(:checked) {\n          background: ", ";\n        }\n\n        &:checked {\n          background: ", ";\n          font-weight: bold;\n        }\n      }\n    }\n\n    &:disabled {\n      background-color: ", ";\n      border-color: ", ";\n      color: ", ";\n\n      &:not([multiple]) > option {\n        background-color: ", ";\n\n        &:checked {\n          opacity: 0.7;\n        }\n      }\n    }\n\n    &:required:not(:valid) {\n      border-color: ", ";\n      ", ";\n    }\n\n    ", "\n  "], ["\n    ", ";\n\n    &:focus {\n      ", ";\n      outline-color: ", ";\n      ", ";\n      ", ";\n      ", ";\n    }\n\n    &[multiple] {\n      > option {\n        &:not(:checked) {\n          background: ", ";\n        }\n\n        &:checked {\n          background: ", ";\n          font-weight: bold;\n        }\n      }\n    }\n\n    &:disabled {\n      background-color: ", ";\n      border-color: ", ";\n      color: ", ";\n\n      &:not([multiple]) > option {\n        background-color: ", ";\n\n        &:checked {\n          opacity: 0.7;\n        }\n      }\n    }\n\n    &:required:not(:valid) {\n      border-color: ", ";\n      ", ";\n    }\n\n    ",
        "\n  "])), placeholder("color: " + lighten(0.5, currentColor) + ";"), outline ? "outline: " + outline : null, outlineColor || (invert ? currentBorderColor : color), outlineOffset ? "outline-offset: " + px(outlineOffset) : null, outlineStyle ? "outline-style: " + outlineStyle : null, outlineWidth ? "outline-width: " + px(outlineWidth) : null, currentBgColor + " linear-gradient(0deg, " + currentBgColor + " 0%, " + currentBgColor + " 100%)", multipleCheckedBgColor + " linear-gradient(0deg, " + multipleCheckedBgColor + " 0%, " + multipleCheckedBgColor + " 100%)", disabledColor, disabledColor, lighten(0.3, currentColor), disabledColor, requiredColor, placeholder("color: " + lighten(0.1, desaturate(0.2, requiredColor)) + ";"), type &&
        "\n      &:read-only {\n        background-color: " + darken(0.02, backgroundColor) + ";\n        color: " + lighten(0.3, currentColor) + ";\n        cursor: not-allowed;\n      }\n    ");
}
export function sx(props) {
    var mergedProps = getProps(props);
    return __assign(__assign({}, systemCSS(props.css)(mergedProps)), responsive(props.responsive)(mergedProps)());
}
export function system(props) {
    return styleSystem(getProps(props));
}
export function variantStyles(props) {
    var dark = props.dark, invert = props.invert, transparent = props.transparent;
    var darkColor = getTheme(props).darkColor;
    var themeColor = getColor(props);
    var backgroundColor = invert ? '#fff' : themeColor;
    var color = invert ? themeColor : getTextColor(props);
    if (transparent) {
        return css(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n      background-color: transparent;\n      border: 0;\n      color: ", ";\n    "], ["\n      background-color: transparent;\n      border: 0;\n      color: ", ";\n    "])), color);
    }
    return css(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n    background-color: ", ";\n    border: ", ";\n    color: ", ";\n  "], ["\n    background-color: ", ";\n    border: ", ";\n    color: ", ";\n  "])), dark ? darkColor : backgroundColor, "1px solid " + (dark ? darkColor : themeColor), dark ? getDarkColor(props) : color);
}
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
