// styled-minimal/esm/utils/theme.js
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { css, keyframes } from 'styled-components';
import { rgba } from './polished';
/**
 * Base
 */
export var breakpoints = ['360px', '400px', '768px', '1024px', '1360px'];
/* eslint-disable prefer-destructuring */
breakpoints.xs = breakpoints[0];
breakpoints.sm = breakpoints[1];
breakpoints.md = breakpoints[2];
breakpoints.lg = breakpoints[3];
breakpoints.xl = breakpoints[4];
export var fontSizes = [10, 12, 14, 16, 20, 26, 32, 40, 48];
export var lineHeight = 1.4;
export var space = [0, 4, 8, 16, 32, 64, 128];
export var headingSizes = Array(6)
    .fill('')
    .map(function (_d, i) { return i + 1; })
    .reduce(function (acc, d) {
    acc["h" + d] = fontSizes[fontSizes.length - d];
    return acc;
}, {});
export var headingWeight = 700;
/**
 * Colors
 */
export var darkColor = '#3C3F41';
export var textColor = '#2E2E2E';
export var borderColor = '#ccc';
export var disabledColor = '#ececec';
export var colors = {
    white: '#fff',
    amber: '#FFC107',
    blue: '#2196F3',
    brown: '#795548',
    cyan: '#00BCD4',
    green: '#4CAF50',
    indigo: '#3F51B5',
    lime: '#CDDC39',
    orange: '#FF9800',
    pink: '#E91E63',
    purple: '#9C27B0',
    red: '#f44336',
    teal: '#009688',
    yellow: '#FFEB3B',
    black: '#000',
    primary: '#4e4141',
    secondary: '#4B515D',
    success: '#00C851',
    warning: '#fda509',
    danger: '#ff4444',
    info: '#33b5e5',
    light: '#f4f4f4',
    dark: '#212121',
    gray03: '#f7f7f7',
    gray05: '#f4f4f4',
    gray10: '#e6e6e6',
    gray20: '#ccc',
    gray30: '#b3b3b3',
    gray40: '#999',
    gray50: '#808080',
    gray60: '#666',
    gray70: '#4d4d4d',
    gray80: '#333',
    gray90: '#1a1a1a',
};
/**
 * Mutation
 */
export var mutation = 0;
export var complexity = 0;
export var seed = 'wabi-sabi';
var baseInput = {
    backgroundColor: '#fff',
    borderColor: borderColor,
    borderRadius: 0,
    borderWidth: 1,
    color: textColor,
    disabledColor: disabledColor,
    fontSize: {
        sm: 14,
        md: 16,
        lg: 20,
    },
    height: {
        sm: 28,
        md: 34,
        lg: 40,
    },
    lineHeight: 1,
    padding: {
        sm: 6,
        md: 8,
        lg: 10,
    },
    requiredColor: '#999',
    validation: {
        invalid: '#f00',
        valid: '#3ac200',
    },
};
var baseCheckboxAndRadio = {
    backgroundColor: '#f7f8f9',
    sizes: {
        sm: 14,
        md: 16,
        lg: 20,
    },
};
var buttonAnimation = keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  0% {\n    background-position: 0 0;\n  }\n\n  100% {\n    background-position: 56px 0;\n  }\n"], ["\n  0% {\n    background-position: 0 0;\n  }\n\n  100% {\n    background-position: 56px 0;\n  }\n"])));
var loader = function (color) {
    if (color === void 0) { color = '#fff'; }
    return css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  animation: ", " 0.5s linear infinite;\n  background-image: repeating-linear-gradient(\n    45deg,\n    ", ",\n    ", " 20px,\n    transparent 20px,\n    transparent 40px\n  );\n  background-size: 56px 56px;\n  pointer-events: none;\n  transition: none !important;\n"], ["\n  animation: ", " 0.5s linear infinite;\n  background-image: repeating-linear-gradient(\n    45deg,\n    ", ",\n    ", " 20px,\n    transparent 20px,\n    transparent 40px\n  );\n  background-size: 56px 56px;\n  pointer-events: none;\n  transition: none !important;\n"])), buttonAnimation, rgba(color, 0.4), rgba(color, 0.4));
};
/**
 * Components
 */
export var alert = {
    borderRadius: 0,
    maxWidth: 450,
    padding: '10px 12px',
};
export var badge = {
    borderRadius: 10,
    fontSize: 12,
    fontWeight: 700,
    lineHeight: 18,
    padding: 5,
    size: 20,
    sizeDot: 10,
};
export var button = {
    borderRadius: {
        xs: 0,
        sm: 0,
        md: 0,
        lg: 0,
        xl: 0,
    },
    disabledOpacity: 0.7,
    fontSize: {
        xs: 12,
        sm: 14,
        md: 16,
        lg: 18,
        xl: 22,
    },
    lineHeight: 1,
    padding: {
        xs: [4, 8],
        sm: [6, 10],
        md: [8, 14],
        lg: [10, 18],
        xl: [12, 20],
    },
    textTransform: 'none',
    loader: loader,
};
export var checkbox = __assign(__assign({}, baseCheckboxAndRadio), { borderRadius: 4, borderColor: borderColor, fontSize: baseInput.fontSize });
export var code = {
    backgroundColor: '#f7f8f9',
    blockColor: '#676767',
    blockPadding: 8,
    border: '1px solid #e5e5e5',
    borderRadius: 0,
    color: '#ff4444',
    fontFamily: 'Courier, monospace',
    padding: 2,
};
export var form = {
    borderColor: borderColor,
    borderRadius: 0,
    borderWidth: 1,
    padding: 10,
};
export var formGroup = {
    borderColor: borderColor,
    borderRadius: 0,
    helpColor: '#999',
    helpMarginTop: space[1],
    inlineMargin: space[2],
    marginBottom: space[3],
    padding: space[2],
};
export var fieldset = {
    borderColor: borderColor,
    borderRadius: 0,
    color: textColor,
    marginBottom: space[3],
    padding: space[2],
};
export var label = {
    color: textColor,
    fontWeight: 'normal',
    inlineFontSize: 14,
    marginBottom: space[1],
};
export var legend = {
    color: textColor,
    fontWeight: 'normal',
    marginBottom: space[2],
};
export var list = {
    borderColor: borderColor,
    borderRadius: 0,
    color: textColor,
    padding: {
        sm: 8,
        md: 12,
        lg: 16,
    },
};
export var input = __assign({}, baseInput);
export var radio = __assign(__assign({}, baseCheckboxAndRadio), { borderRadius: '50%', borderColor: borderColor, fontSize: baseInput.fontSize });
export var select = __assign(__assign({}, baseInput), { checkedBgColor: '' });
export var textarea = __assign(__assign({}, baseInput), { lineHeight: 1.4 });
export var switchOptions = {
    innerBorderRadius: '50%',
    marginLabel: 8,
    sizes: {
        xs: {
            borderRadius: 7,
            height: 14,
            space: 2,
            width: 24,
        },
        sm: {
            borderRadius: 10,
            height: 20,
            space: 3,
            width: 35,
        },
        md: {
            borderRadius: 14,
            height: 26,
            space: 3,
            width: 47,
        },
        lg: {
            borderRadius: 16,
            height: 32,
            space: 3,
            width: 57,
        },
        xl: {
            borderRadius: 20,
            height: 40,
            space: 3,
            width: 74,
        },
    },
};
export var table = {
    borderColors: { primary: '#ccc', secondary: '#32383e' },
    captionColor: '#ccc',
    captionPadding: 6,
    colors: { primary: '#fff', secondary: '#000' },
    headColors: { light: '#f4f4f4', dark: '#000' },
    padding: {
        sm: 6,
        md: 12,
        lg: 18,
    },
    stripedColors: { primary: '#ededed', secondary: '#262626' },
};
var templateObject_1, templateObject_2;
