// styled-minimal/lib/utils/helpers.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.whitelist = exports.spacer = exports.replaceShorthands = exports.responsive = exports.parseNumericCSSProp = exports.mergeTheme = exports.px = exports.pick = exports.limitNumber = exports.isString = exports.isNumericString = exports.isNumber = exports.isDefined = exports.isCSSUnit = exports.isBoolean = exports.getYiq = exports.getTheme = exports.getTextColor = exports.getStyles = exports.getReadableColor = exports.getRandomInt = exports.getProps = exports.getPercentage = exports.getMutation = exports.getMediaQueries = exports.getAccentColor = exports.getDarkColor = exports.getColor = exports.createMediaQuery = void 0;
/*eslint-disable @typescript-eslint/no-use-before-define */
var deepmerge = require("deepmerge");
var css_1 = require("@styled-system/css");
var polished_1 = require("./polished");
var styles_1 = require("./styles");
var defaultTheme = require("./theme");
/**
 * Generate the media query
 */
function createMediaQuery(size, mediaQueries) {
    if (isCSSUnit(size) || isNumericString(size)) {
        return "@media screen and (min-width: " + px(size) + ")";
    }
    return mediaQueries[size];
}
exports.createMediaQuery = createMediaQuery;
/**
 * Get color from theme
 */
function getColor(props, _a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.base, base = _c === void 0 ? 'primary' : _c, _d = _b.includeColor, includeColor = _d === void 0 ? false : _d;
    var _e = props.variant, variant = _e === void 0 ? '' : _e;
    var colors = getTheme(props).colors;
    try {
        if (!variant && includeColor) {
            var color = getStyles(props).color;
            if (color) {
                return color;
            }
        }
        return colors[variant] || colors[base];
    }
    catch (error) {
        return colors.primary;
    }
}
exports.getColor = getColor;
function getDarkColor(props) {
    var darkColor = getTheme(props).darkColor;
    var color = getColor(props);
    return getReadableColor(darkColor, color);
}
exports.getDarkColor = getDarkColor;
function getAccentColor(color, lighter) {
    if (lighter === void 0) { lighter = false; }
    try {
        var luminance = polished_1.getLuminance(color);
        var accent = polished_1.darken(0.1, color);
        if (lighter) {
            accent = polished_1.lighten(0.1, color);
            if (luminance < 0.1) {
                accent = polished_1.lighten(0.2, color);
            }
            else if (luminance > 0.8) {
                accent = polished_1.darken(0.05, color);
            }
            else if (luminance > 0.6) {
                accent = polished_1.lighten(0.05, color);
            }
            return accent;
        }
        if (luminance < 0.05) {
            accent = polished_1.lighten(0.1, color);
        }
        else if (luminance < 0.1) {
            accent = polished_1.darken(0.1, color);
        }
        return accent;
    }
    catch (error) {
        /* istanbul ignore next */
        if (process.env.NODE_ENV === 'development') {
            console.error(error); //eslint-disable-line no-console
        }
        return color;
    }
}
exports.getAccentColor = getAccentColor;
function getMediaQueries(breakpoints) {
    return Object.keys(breakpoints)
        .filter(function (d) { return Number.isNaN(parseInt(d, 10)); })
        .reduce(function (acc, d) {
        acc[d] = "@media screen and (min-width: " + px(breakpoints[d]) + ")";
        return acc;
    }, {});
}
exports.getMediaQueries = getMediaQueries;
function getMutation(props, skipTheme) {
    var _a, _b, _c;
    var theme = getTheme(props);
    var styles = whitelist(getStyles(props), ['margin', 'border']);
    var complexity = (_a = props.complexity) !== null && _a !== void 0 ? _a : (!skipTheme ? theme.complexity : 0);
    var mutation = (_b = props.mutation) !== null && _b !== void 0 ? _b : (!skipTheme ? theme.mutation : 0);
    var seed = (_c = props.seed) !== null && _c !== void 0 ? _c : theme.seed;
    if (Array.isArray(mutation)) {
        var rate = mutation[0], width = mutation[1], height = mutation[2];
        return __assign({ complexity: complexity, height: height, mutation: rate, seed: seed, width: width }, styles);
    }
    return __assign({ complexity: complexity, mutation: mutation, seed: seed }, styles);
}
exports.getMutation = getMutation;
/**
 * Get percentage of positive numbers
 */
function getPercentage(input, total) {
    return Number(parseFloat(String((100 * Math.max(0, input)) / Math.abs(total))).toFixed(2));
}
exports.getPercentage = getPercentage;
function getProps(props) {
    if (props === void 0) { props = {}; }
    return __assign(__assign({}, props), { theme: mergeTheme(props.theme) });
}
exports.getProps = getProps;
function getRandomInt(min, max, random) {
    if (min === void 0) { min = 1; }
    if (max === void 0) { max = 10; }
    if (random === void 0) { random = Math.random(); }
    return Math.floor(random * (max - min + 1) + min);
}
exports.getRandomInt = getRandomInt;
function getReadableColor(baseColor, color) {
    var limit = 75;
    var colorDiff = Math.abs(getYiq(baseColor) - getYiq(color));
    var readableColor = color;
    if (colorDiff >= limit) {
        return readableColor;
    }
    while (colorDiff < limit) {
        readableColor = polished_1.lighten(0.05, readableColor);
        colorDiff = Math.abs(getYiq(baseColor) - getYiq(readableColor));
    }
    return readableColor;
}
exports.getReadableColor = getReadableColor;
function getStyles(props) {
    var theme = getTheme(props);
    return replaceShorthands(css_1.default(pick(props))({ theme: theme }));
}
exports.getStyles = getStyles;
function getTextColor(props) {
    try {
        var colors = getTheme(props, 'colors');
        var currentColor = getColor(props);
        return getYiq(currentColor) > 135 ? colors.black : colors.white;
    }
    catch (error) {
        return '#000';
    }
}
exports.getTextColor = getTextColor;
/**
 * Get the merged theme
 */
function getTheme(props, path) {
    var theme = mergeTheme((props === null || props === void 0 ? void 0 : props.theme) || {});
    return path ? theme[path] : theme;
}
exports.getTheme = getTheme;
/**
 * Color contrast
 */
function getYiq(color) {
    try {
        var _a = polished_1.parseToRgb(color), red = _a.red, green = _a.green, blue = _a.blue;
        return (red * 299 + green * 587 + blue * 114) / 1000;
    }
    catch (error) {
        return 0;
    }
}
exports.getYiq = getYiq;
/** Check if a variable is a boolean **/
function isBoolean(value) {
    return typeof value === 'boolean';
}
exports.isBoolean = isBoolean;
function isCSSUnit(value) {
    var units = ['em', 'px', 'rem', 'vh', 'vmax', 'vmin', 'vw'];
    var regex = new RegExp("\\d+(" + units.join('|') + ")$");
    return typeof value === 'string' && regex.test(value);
}
exports.isCSSUnit = isCSSUnit;
/** Check if a variable is defined **/
function isDefined(value) {
    return typeof value !== 'undefined';
}
exports.isDefined = isDefined;
/** Check if a variable is a number **/
function isNumber(value) {
    return typeof value === 'number';
}
exports.isNumber = isNumber;
function isNumericString(value) {
    return isString(value) && value.length > 0 && !Number.isNaN(Number(value));
}
exports.isNumericString = isNumericString;
function isString(value) {
    return typeof value === 'string';
}
exports.isString = isString;
function limitNumber(num, min, max) {
    if (min === void 0) { min = 0; }
    if (max === void 0) { max = 100; }
    return Math.min(Math.max(num, min), max);
}
exports.limitNumber = limitNumber;
function pick(props) {
    var regex = new RegExp("^(" + styles_1.styleSystem.propNames.join('|') + ")$");
    var rest = {};
    for (var key in props) {
        if ({}.hasOwnProperty.call(props, key) && regex.test(key)) {
            rest[key] = props[key];
        }
    }
    return rest;
}
exports.pick = pick;
/** Get textual unit value */
function px(value) {
    return isNumber(value) || isNumericString(value) ? value + "px" : value || '0px';
}
exports.px = px;
/**
 * Merge ThemeProvides's theme with the default theme
 */
function mergeTheme(theme) {
    if (theme === void 0) { theme = {}; }
    return deepmerge(defaultTheme, theme, {
        arrayMerge: function (_dest, source) { return source; },
    });
}
exports.mergeTheme = mergeTheme;
function parseNumericCSSProp(value) {
    if (isNumber(value)) {
        return value;
    }
    if (value.indexOf('px') > 0) {
        return parseInt(value, 10);
    }
    var baseFontSize = parseFloat(getComputedStyle(document.documentElement).fontSize);
    return parseFloat(value) * baseFontSize;
}
exports.parseNumericCSSProp = parseNumericCSSProp;
/**
 * SC Helper to generate responsive media queries1
 */
function responsive(rules) {
    return function (props) {
        var breakpoints = getTheme(props).breakpoints;
        var entries = {};
        var mediaQueries = getMediaQueries(breakpoints);
        for (var rule in rules) {
            /* istanbul ignore else */
            if ({}.hasOwnProperty.call(rules, rule)) {
                var breakpoint = rule;
                var styles = css_1.default(replaceShorthands(rules[breakpoint]))(props);
                var query = createMediaQuery(breakpoint, mediaQueries);
                if (breakpoint === '_') {
                    Object.entries(styles).forEach(function (_a) {
                        var k = _a[0], v = _a[1];
                        entries[k] = v;
                    });
                }
                else if (query) {
                    entries[query] = styles;
                }
            }
        }
        return css_1.default(entries);
    };
}
exports.responsive = responsive;
function replaceShorthands(styles) {
    return Object.entries(styles).reduce(function (acc, _a) {
        var k = _a[0], v = _a[1];
        switch (k) {
            case 'cl': {
                acc.color = v;
                break;
            }
            case 'h': {
                acc.height = v;
                break;
            }
            case 'w': {
                acc.width = v;
                break;
            }
            default: {
                acc[k] = v;
                break;
            }
        }
        return acc;
    }, {});
}
exports.replaceShorthands = replaceShorthands;
/**
 * Helper to get the corresponding item from the space scale
 */
function spacer(value) {
    return function (props) {
        var space = getTheme(props).space;
        return px(space[value] || value);
    };
}
exports.spacer = spacer;
function whitelist(input, entries) {
    return Object.entries(input).reduce(function (acc, _a) {
        var key = _a[0], value = _a[1];
        if (entries.some(function (e) { return key.startsWith(e); })) {
            acc[key] = value;
        }
        return acc;
    }, {});
}
exports.whitelist = whitelist;
