// styled-minimal/lib/utils/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.styleSystem = exports.transformProps = exports.textTransform = exports.textDecoration = exports.outlines = exports.filter = exports.cursor = exports.boxShadow = exports.animation = void 0;
var styled_system_1 = require("styled-system");
var layout = styled_system_1.compose(styled_system_1.display, styled_system_1.height, styled_system_1.maxHeight, styled_system_1.maxWidth, styled_system_1.minHeight, styled_system_1.minWidth, styled_system_1.overflow, styled_system_1.overflowX, styled_system_1.overflowY, styled_system_1.verticalAlign, styled_system_1.width);
exports.animation = styled_system_1.system({
    animation: true,
    animationDelay: true,
    animationDirection: true,
    animationDuration: true,
    animationFillMode: true,
    animationIterationCount: true,
    animationName: true,
    animationPlayState: true,
    animationTimingFunction: true,
});
exports.boxShadow = styled_system_1.system({
    boxShadow: true,
});
exports.cursor = styled_system_1.system({
    cursor: true,
});
exports.filter = styled_system_1.system({
    filter: true,
});
exports.outlines = styled_system_1.system({
    outline: true,
    outlineColor: true,
    outlineOffset: true,
    outlineStyle: true,
    outlineWidth: true,
});
var shorthands = styled_system_1.system({
    cl: {
        property: 'color',
        scale: 'colors',
    },
    h: {
        property: 'height',
    },
    w: {
        property: 'height',
    },
});
exports.textDecoration = styled_system_1.system({
    textDecoration: true,
});
exports.textTransform = styled_system_1.system({
    textTransform: true,
});
exports.transformProps = styled_system_1.system({
    backfaceVisibility: true,
    perspective: true,
    perspectiveOrigin: true,
    rotate: true,
    scale: true,
    transform: true,
    transformOrigin: true,
    transition: true,
});
exports.styleSystem = styled_system_1.compose(exports.animation, styled_system_1.background, styled_system_1.border, exports.boxShadow, styled_system_1.color, exports.cursor, exports.filter, styled_system_1.flexbox, styled_system_1.grid, layout, exports.outlines, styled_system_1.position, shorthands, styled_system_1.space, styled_system_1.typography, exports.textDecoration, exports.textTransform, exports.transformProps);
