// @gilbarbara/helpers/lib/numbers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.round = exports.rangeLimit = exports.randomInt = exports.pad = exports.floor = exports.ceil = void 0;
/**
 * Ceil decimal numbers
 */
function ceil(input, digits) {
    if (digits === void 0) { digits = 2; }
    var factor = Math.pow(10, digits);
    return Math.ceil(input * factor) / factor;
}
exports.ceil = ceil;
/**
 * Floor decimal numbers
 */
function floor(input, digits) {
    if (digits === void 0) { digits = 2; }
    var factor = Math.pow(10, digits);
    return Math.floor(input * factor) / factor;
}
exports.floor = floor;
/**
 * Pad a number with zeros
 */
function pad(input, length) {
    if (length === void 0) { length = 2; }
    return "".concat(input).padStart(length, '0');
}
exports.pad = pad;
/**
 * Returns a random integer
 */
function randomInt(min, max) {
    if (min === void 0) { min = 0; }
    if (max === void 0) { max = 10; }
    if (min >= max) {
        return max;
    }
    return Math.floor(Math.random() * (max - min + 1) + min);
}
exports.randomInt = randomInt;
/**
 * Limit number between range
 */
function rangeLimit(value, min, max) {
    if (min === void 0) { min = 0; }
    if (max === void 0) { max = 100; }
    return Math.min(Math.max(value, min), max);
}
exports.rangeLimit = rangeLimit;
/**
 * Round decimal numbers
 */
function round(input, digits) {
    if (digits === void 0) { digits = 2; }
    var factor = Math.pow(10, digits);
    return Math.round(input * factor) / factor;
}
exports.round = round;
